Force people to call from_meta_and_device directly

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87943
* #87911
* __->__ #87903

It was pretty hard to tell at call site if I was doing device meta
convert or not.  This gets rid of the "dual" API and forces people
to call the method manually for the device case.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Force people to call from_meta_and_device directly

It was pretty hard to tell at call site if I was doing device meta
convert or not.  This gets rid of the "dual" API and forces people
to call the method manually for the device case.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -318,7 +312,8 @@ def non_kwarg_to(fake_mode, func, *args, **kwargs):
     inp = new_kwargs.pop("input")
     r = func(inp, **new_kwargs)
-    return fake_mode.fake_tensor_converter(fake_mode, r, out_device)
+    # TODO: I think this does the wrong thing if r is inp
```
these are guaranteed copies, i thought. this is post-composite implicit. no-op aten::to shouldn't invoke `__torch_dispatch__` right ?

eellison replied:
If my comment is correct can you remove `# TODO: I think this does the wrong thing if r is inp` comment before landing

ezyang replied:
You registered this implementation to `aten.to_device` which is a composite

```
- func: to.device(Tensor(a) self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)
  variants: method
  device_check: NoCheck
  device_guard: False
```

as its alias annotations say, it can return an alias of the input tensor.

albanD replied:
Indeed, `.to()` can return `self` if nothing is being done.
Should this be:
```python
if r is inp:
  return r
else:
  return fake_mode....
```

eellison replied:
Got it. I didn't think composites could even be hit in `__torch__dispatch__`, so I guess I didn't check alias carefully. Is that not a bug ?

ezyang replied:
Well yeah, my suspicion is that override is dead. Need to test.

albanD replied:
Well, if the converter is a no-op for already fake Tensor, that should be fine.

eellison reviewed (approved):
thanks! yea this is clearer

ezyang committed:
Update on "Force people to call from_meta_and_device directly"


It was pretty hard to tell at call site if I was doing device meta
convert or not.  This gets rid of the "dual" API and forces people
to call the method manually for the device case.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Force people to call from_meta_and_device directly"


It was pretty hard to tell at call site if I was doing device meta
convert or not.  This gets rid of the "dual" API and forces people
to call the method manually for the device case.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Force people to call from_meta_and_device directly"


It was pretty hard to tell at call site if I was doing device meta
convert or not.  This gets rid of the "dual" API and forces people
to call the method manually for the device case.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]
